lib/Sema/TypeCheckType.cpp

```diff
@@ -843,11 +843,36 @@ static Type diagnoseUnknownType(TypeChecker &tc, DeclContext *dc,
     tc.diagnose(comp->getIdLoc(), diag::no_module_type,
                 comp->getIdentifier(), moduleType->getModule()->getName());
   } else {
-    tc.diagnose(comp->getIdLoc(), diag::invalid_member_type,
-                comp->getIdentifier(), parentType)
-      .highlight(parentRange);
-  }
+    // Situation where class tries to inherit from itself, such
+    // would produce an assertion when trying to lookup members of the class.
+    auto resolver = tc.Context.getLazyResolver();
+    if (auto superClass = parentType->getSuperclass(resolver)) {
+      if (superClass->isEqual(parentType)) {
+        auto decl = parentType->castTo<NominalType>()->getDecl();
+        tc.diagnose(decl->getLoc(), diag::circular_class_inheritance,
+                    decl->getNameStr());
+        return ErrorType::get(tc.Context);
+      }
+    }
+
+    LookupResult memberLookup;
+    // Let's try to lookup given identifier as a member of the parent type,
+    // this allows for more precise diagnostic, which distinguishes between
+    // identifier not found as a member type vs. not found at all.
+    memberLookup = tc.lookupMember(dc, parentType, comp->getIdentifier(),
+                                   NameLookupFlags::IgnoreAccessibility);
```

Comment:
I'm concerned about dropping all other lookup options. You should also use KnownPrivate so that this doesn't get recorded as a dependency.